#### RandNet

In [ ]:
!java -jar RankLib-patched.jar -train Fold1/train.txt -test Fold1/test.txt -validate Fold1/vali.txt -ranker 1 -metric2t NDCG@10 -metric2T NDCG@8 -save RankNet.txt

96      | 0.106          | 0.4871    | 0.5104    | 
97      | 0.1061         | 0.4871    | 0.5104    | 
98      | 0.1061         | 0.4875    | 0.5104    | 
99      | 0.106          | 0.4874    | 0.5107    | 
100     | 0.1061         | 0.4873    | 0.5107    | 
--------------------------------------------------
Finished sucessfully.
NDCG@10 on training data: 0.4889
NDCG@10 on validation data: 0.5163
---------------------------------
NDCG@8 on test data: 0.4668

Model saved to: RankNet.txt

#### ListNet

In [ ]:
!java -jar RankLib-patched.jar -train Fold1/train.txt -test Fold1/test.txt -validate Fold1/vali.txt -ranker 7 -metric2t NDCG@10 -metric2T NDCG@8 -save ListNet.txt

1495    | 0.0            | 0.4857    | 0.514     | 
1496    | 0.0            | 0.4857    | 0.514     | 
1497    | 0.0            | 0.4857    | 0.514     | 
1498    | 0.0            | 0.4857    | 0.514     | 
1499    | 0.0            | 0.4857    | 0.514     | 
1500    | 0.0            | 0.4857    | 0.514     | 
--------------------------------------------------
Finished sucessfully.
NDCG@10 on training data: 0.4848
NDCG@10 on validation data: 0.5143
---------------------------------
NDCG@8 on test data: 0.4556

Model saved to: ListNet.txt

#### LambdaMart

In [ ]:
!java -jar RankLib-patched.jar -train Fold1/train.txt -test Fold1/test.txt -validate Fold1/vali.txt -ranker 6 -metric2t NDCG@10 -metric2T NDCG@8 -save LambdaMart.txt

130     | 0.6003    | 0.5416    | 
131     | 0.6006    | 0.5412    | 
132     | 0.6002    | 0.5396    | 
133     | 0.6003    | 0.5436    | 
134     | 0.6005    | 0.5443    | 
135     | 0.6008    | 0.5435    | 
---------------------------------
Finished sucessfully.
NDCG@10 on training data: 0.5322
NDCG@10 on validation data: 0.5477
---------------------------------
NDCG@8 on test data: 0.4858

Model saved to: LambdaMart.txt

#### 多个模型进行对比

##### 生成baseline NDCG数据

In [11]:
!java -jar RankLib-patched.jar -test Fold1/test.txt -metric2T NDCG@10 -idv output/baseline.ndcg.txt


[+] General Parameters:
Model file:	
Feature normalization: No
Test metric:	NDCG@10

Reading feature file [Fold1/test.txt]: 0... 
Reading feature file [Fold1/test.txt]... [Done.]            
(156 ranked lists, 2874 entries read)
NDCG@10 on test data: 0.3257
Per-ranked list performance saved to: output/baseline.ndcg.txt


##### 分别训练三个模型（RankNet, ListNet, LambdaMart）

In [ ]:
# RankNet
!java -jar RankLib-patched.jar -train Fold1/train.txt -test Fold1/test.txt -validate Fold1/vali.txt -ranker 1 -metric2t NDCG@10 -metric2T NDCG@10 -save model_ranknet.txt
# ListNet
!java -jar RankLib-patched.jar -train Fold1/train.txt -test Fold1/test.txt -validate Fold1/vali.txt -ranker 7 -metric2t NDCG@10 -metric2T NDCG@10 -save model_listnet.txt
# LambdaMart
!java -jar RankLib-patched.jar -train Fold1/train.txt -test Fold1/test.txt -validate Fold1/vali.txt -ranker 6 -metric2t NDCG@10 -metric2T NDCG@10 -save model_lambdamart.txt


##### 得出三个模型的NDCG结果（RankNet, ListNet, LambdaMart）

In [13]:
#RankNet
!java -jar RankLib-patched.jar -load model_ranknet.txt -test Fold1/test.txt -metric2T NDCG@10 -idv output/ranknet.ndcg.txt
#ListNet
!java -jar RankLib-patched.jar -load model_listnet.txt -test Fold1/test.txt -metric2T NDCG@10 -idv output/listnet.ndcg.txt
#LambdaMart
!java -jar RankLib-patched.jar -load model_lambdamart.txt -test Fold1/test.txt -metric2T NDCG@10 -idv output/lambdamart.ndcg.txt


[+] General Parameters:
Model file:	model_ranknet.txt
Feature normalization: No
Test metric:	NDCG@10
Model:		RankNet

Reading feature file [Fold1/test.txt]: 0... 
Reading feature file [Fold1/test.txt]... [Done.]            
(156 ranked lists, 2874 entries read)
NDCG@10 on test data: 0.4814
Per-ranked list performance saved to: output/ranknet.ndcg.txt

[+] General Parameters:
Model file:	model_listnet.txt
Feature normalization: No
Test metric:	NDCG@10
Model:		ListNet

Reading feature file [Fold1/test.txt]: 0... 
Reading feature file [Fold1/test.txt]... [Done.]            
(156 ranked lists, 2874 entries read)
NDCG@10 on test data: 0.4685
Per-ranked list performance saved to: output/listnet.ndcg.txt

[+] General Parameters:
Model file:	model_lambdamart.txt
Feature normalization: No
Test metric:	NDCG@10
Model:		LambdaMART

Reading feature file [Fold1/test.txt]: 0... 
Reading feature file [Fold1/test.txt]... [Done.]            
(156 ranked lists, 2874 entries read)
NDCG@10 on test data: 0

##### 多个模型对比（RankNet, ListNet, LambdaMart）

In [15]:
!java -cp RankLib-patched.jar ciir.umass.edu.eval.Analyzer -all output/ -base baseline.ndcg.txt > analysis.txt

##### analysis.txt

Reading output/baseline.ndcg.txt... 157 ranked lists [Done]
Reading output/lambdamart.ndcg.txt... 157 ranked lists [Done]
Reading output/listnet.ndcg.txt... 157 ranked lists [Done]
Reading output/ranknet.ndcg.txt... 157 ranked lists [Done]


Overall comparison
------------------------------------------------------------------------
System	Performance	Improvement	Win	Loss	p-value
baseline.ndcg.txt [baseline]	0.3257
lambdamart.ndcg.txt	0.4891	+0.1633 (+50.15%)	79	25	0.0
listnet.ndcg.txt	0.4685	+0.1428 (+43.84%)	80	21	0.0
ranknet.ndcg.txt	0.4814	+0.1557 (+47.79%)	82	19	0.0


Detailed break down
------------------------------------------------------------------------
	[ < -100%)	[-100%, -75%)	[-75%, -50%)	[-50%, -25%)	[-25%, 0%)	(0%, +25%]	(+25%, +50%]	(+50%, +75%]	(+75%, +100%]	( > +100%]
lambdamart.ndcg.txt	0	0	0	6	19	24	35	15	5	0
listnet.ndcg.txt	0	0	0	5	16	34	29	15	2	0
ranknet.ndcg.txt	0	0	0	5	14	33	29	18	2	0
